### PR TITLE
Connector orchestration

### PR DESCRIPTION
**Pull request by ryzhyk**
Implements RFC #2096.

See individual commits for details.

TODOs:
* Rework pipeline manager API as part of ongoing API refactoring
* Documentation (post-refactoring)

Is this a user-visible change (yes/no): yes

### CODE REVIEW

File: crates/adapters/src/controller/stats.rs [hunk 1]
```diff
@@ -846,16 +866,29 @@ pub struct InputEndpointStatus {
 
     /// Whether this input endpoint is [fault tolerant](crate#fault-tolerance).
     pub is_fault_tolerant: bool,
+
+    /// Endpoint has been paused by the user.
+    ///
+    /// When `true`, the endpoint doesn't produce any data even when the pipeline
+    /// is running.
+    ///
+    /// This flag is set to `true` on startup if the `pause_on_startup` flag in the
```
Comment (lalithsuresh) [inline crates/adapters/src/controller/stats.rs:875]: As an end user, a desired state of "paused: true/false" in the connector state would be easier to think about than a "paused_on_startup: true/false".

**Review by lalithsuresh — COMMENTED** [review]
As a user, I'd prefer to only state the desired state for the connector ("desired_state: running/paused", "current_state: <whatever runtime state is>"). A "paused_at_startup" flag instead requires two different APIs: one to make the connector start in the paused state and another to switch it to the paused state at runtime.

It would seem that internally, the endpoint can be in one of three states: running, backpressure, or paused. I think this internal state is currently called desired state?

**Comment by ryzhyk** [comment]
> It's a start to get rolling. But as a user, I'd prefer to only state the desired state for the connector ("desired_state: running/paused", "current_state: whatever-runtime-state-is"). A "paused_at_startup" flag instead requires two different APIs: one to make the connector start in the paused state and another to switch it to the paused state at runtime.
> 
> It would seem that internally, the endpoint can be in one of three states: running, paused-by-pipeline / paused-backpressure, or paused-by-user. I think this internal state is currently called desired state?

We need two separate flags with two different lifetimes. `paused_on_startup` is part of connector configuration, which exists before the pipeline is created and never changes.  You can think of it as desired initial state of the connector. I made it a boolean rather than an enum, since that seemed more user friendly, but it can also be an enum with two possible values `paused`/`running`.

The second flag is the runtime flag. I called it `paused_by_user`, but it could be simply `paused`. Or it could be an enum as well.

**Comment by lalithsuresh** [comment]
> > It's a start to get rolling. But as a user, I'd prefer to only state the desired state for the connector ("desired_state: running/paused", "current_state: whatever-runtime-state-is"). A "paused_at_startup" flag instead requires two different APIs: one to make the connector start in the paused state and another to switch it to the paused state at runtime.
> > It would seem that internally, the endpoint can be in one of three states: running, paused-by-pipeline / paused-backpressure, or paused-by-user. I think this internal state is currently called desired state?
> 
> We need two separate flags with two different lifetimes. `paused_on_startup` is part of connector configuration, which exists before the pipeline is created and never changes. You can think of it as desired initial state of the connector. I made it a boolean rather than an enum, since that seemed more user friendly, but it can also be an enum with two possible values `paused`/`running`.
> 
> The second flag is the runtime flag. I called it `paused_by_user`, but it could be simply `paused`. Or it could be an enum as well.

Yes, so what I'm questioning is the two lifetimes. Both could be one desired state field, called "paused", which is less confusing for me as a user. 

When I start a pipeline that has a connector with desired state `paused=true`, it has the effect of "the connector is paused on startup". When I changed desired state paused=false afterwards at runtime, it's equivalent to the "paused by user" scenario. Seems like a simpler API and more in line with what I see in other cloud APIs.

**Comment by mihaibudiu** [comment]
Why isn't everything paused on startup?
You don't want to change sql on backfill

**Comment by ryzhyk** [comment]
> > > It's a start to get rolling. But as a user, I'd prefer to only state the desired state for the connector ("desired_state: running/paused", "current_state: whatever-runtime-state-is"). A "paused_at_startup" flag instead requires two different APIs: one to make the connector start in the paused state and another to switch it to the paused state at runtime.
> > > It would seem that internally, the endpoint can be in one of three states: running, paused-by-pipeline / paused-backpressure, or paused-by-user. I think this internal state is currently called desired state?
> > 
> > 
> > We need two separate flags with two different lifetimes. `paused_on_startup` is part of connector configuration, which exists before the pipeline is created and never changes. You can think of it as desired initial state of the connector. I made it a boolean rather than an enum, since that seemed more user friendly, but it can also be an enum with two possible values `paused`/`running`.
> > The second flag is the runtime flag. I called it `paused_by_user`, but it could be simply `paused`. Or it could be an enum as well.
> 
> Yes, so what I'm questioning is the two lifetimes. Both could be one desired state field, called "paused", which is less confusing for me as a user.
> 
> When I start a pipeline that has a connector with desired state `paused=true`, it has the effect of "the connector is paused on startup". If I change desired state to `paused=false` later, it's equivalent to the "paused/unpaused by user" scenario. Seems like a simpler API and more in line with what I see in other cloud APIs.
> 
> In k8s for example, if I say "spec.replicas=5" for a deployment, that's sufficient to express the initial and later desired states.

Where would you put this one desired state field?

**Comment by ryzhyk** [comment]
> Why isn't everything paused on startup? You don't want to change sql on backfill

Currently you can create a pipeline in either paused or running state. In the former case it will create all connectors in paused state. In the latter case it will unpause all connectors.

We are adding a new mechanism that allows the user to control the state of individual connectors via the API even when the pipeline is running. This doesn't require changing SQL at runtime.

**Comment by lalithsuresh** [comment]
> > Yes, so what I'm questioning is the two lifetimes. Both could be one desired state field, called "paused", which is less confusing for me as a user.
> 
> Where would you put this one desired state field?

I'm thinking `ConnectorConfig`, which describes the overall spec of the connector and is used to initialize the connector.

Then again in `PipelineRuntimeState` which has `desired_state` and `current_state` for the whole pipeline, but in addition will have a similar pair for each connector it has.

**Comment by ryzhyk** [comment]
> > > Yes, so what I'm questioning is the two lifetimes. Both could be one desired state field, called "paused", which is less confusing for me as a user.
> > 
> > 
> > Where would you put this one desired state field?
> 
> I'm thinking `ConnectorConfig`, which describes the overall spec of the connector and is used to initialize the connector.
> 
> Then again in `PipelineRuntimeState` which has `desired_state` and `current_state` for the whole pipeline, but in addition will have a similar pair for each connector it has.

This is precisely what I did. Shall I just rename the fields to something else?

**Comment by lalithsuresh** [comment]
> > > > Yes, so what I'm questioning is the two lifetimes. Both could be one desired state field, called "paused", which is less confusing for me as a user.
> > > 
> > > 
> > > Where would you put this one desired state field?
> > 
> > 
> > I'm thinking `ConnectorConfig`, which describes the overall spec of the connector and is used to initialize the connector.
> > Then again in `PipelineRuntimeState` which has `desired_state` and `current_state` for the whole pipeline, but in addition will have a similar pair for each connector it has.
> 
> This is precisely what I did. Shall I just rename the fields to something else?

Ah, then yes, let's just rename it to be consistent with the terms used for PipelineRuntimeState too (desired_state: paused/running, current_state: paused/running).

**Comment by ryzhyk** [comment]
There is no need to distinguish desired and actual state for connectors. Unlike with pipelines, the state transition happens instantly, so a single flag does the job just fine.

**Review by snkas — COMMENTED** [review]
In general it looks good, starting and pausing existing endpoints/connectors of tables. 

Regarding the naming of:
"endpoints" vs. "input_endpoints" vs. "output_endpoints" vs. "connectors" vs. "input_connectors" vs. "output_connectors", and the table or view namespace where they exist

We should  be consistent throughout what things are named and how they are referred to. E.g., if we name it endpoint in the API, then everywhere in documentation, code and API it should state "Create a Kafka input endpoint for table <example>" etc..

Will there be output_endpoints as well later on?

**Comment by ryzhyk** [comment]
> In general it looks good, starting and pausing existing endpoints/connectors of tables.
> 
> Regarding the naming of: "endpoints" vs. "input_endpoints" vs. "output_endpoints" vs. "connectors" vs. "input_connectors" vs. "output_connectors", and the table or view namespace where they exist
> 
> We should be consistent throughout what things are named and how they are referred to. E.g., if we name it endpoint in the API, then everywhere in documentation, code and API it should state "Create a Kafka input endpoint for table " etc..
> 
> Will there be output_endpoints as well later on?

@snkas , right now we have two different concepts (connectors and endpoints). This won't be necessary with the new API, so we will be able to reduce the number of concepts to one and rename the API endpoint. Regarding the most appropriate name, it can probably be one of "connector" or "source", but definitely not "endpoint".

**Pull request merged**